bin/sessionmirror.py

```diff
@@ -0,0 +1,749 @@
+    print(xml_path)
+
+    sess_type = src_sess.datatype()
+    dst_sess.create(experiments=sess_type)
+    copy_attributes(src_sess, dst_sess)
+
+    # Process each scan of session
+    for src_scan in src_sess.scans().fetchall('obj'):
+        scan_label = src_scan.label()
+
+        print('INFO:Processing scan:%s...' % scan_label)
+        dst_scan = dst_sess.scan(scan_label)
+        scan_cache_dir = op.join(sess_cache_dir, scan_label)
+        copy_scan(src_scan, dst_scan, scan_cache_dir)
+
+    # Process each assessor of session
+    for src_assr in src_sess.assessors():
+        assr_label = src_assr.label()
+        print('INFO:Processing assessor:%s:...' % assr_label)
+        dst_assr = dst_sess.assessor(assr_label)
+        assr_cache_dir = op.join(sess_cache_dir, assr_label)
+        copy_assr(src_assr, dst_assr, assr_cache_dir)
+
+    for src_res in src_sess.resources().fetchall('obj'):
+        res_label = src_res.label()
+
+        print('INFO:Processing resource:%s...' % (res_label))
+
+        dst_res = dst_sess.resource(res_label)
+
+        res_cache_dir = op.join(scan_cache_dir, res_label)
+
+        copy_res(src_res, dst_res, res_cache_dir, use_zip=True)
+
+def copy_scan(src_scan, dst_scan, scan_cache_dir):
+    '''Copy scan from source XNAT to destination XNAT'''
+
+    scan_type = src_scan.datatype()
+    if scan_type == '':
+             scan_type = 'xnat:otherDicomScanData'
+    dst_scan.create(scans=scan_type)
+    copy_attributes(src_scan, dst_scan)
+
+    # Process each resource of scan
+    for src_res in src_scan.resources().fetchall('obj'):
+        res_label = src_res.label()
+
+        print('INFO:Processing resource:%s...' % (res_label))
+
+        dst_res = dst_scan.resource(res_label)
+
+        res_cache_dir = op.join(scan_cache_dir, res_label)
+        if res_label == 'SNAPSHOTS':
+            copy_res(src_res, dst_res, res_cache_dir)
+        else:
+            copy_res(src_res, dst_res, res_cache_dir, use_zip=True)
+
+
+def copy_res(src_res, dst_res, res_cache_dir, use_zip=False):
+    '''Copy resource from source XNAT to destination XNAT'''
+    # Create cache dir
+    if not op.exists(res_cache_dir):
+        os.makedirs(res_cache_dir)
+
+    # Prepare resource and check for empty
+    is_empty = False
+    print(dst_res._uri)
+    if not dst_res.exists():
+        dst_res.create()
+        is_empty = True
+    elif is_empty_resource(dst_res):
+        is_empty = True
+
+    # Check for empty source
+    if is_empty_resource(src_res):
+        print('WARN:empty resource, nothing to copy')
+        return
+
+    if is_empty:
+        if use_zip:
+            # Try to copy as zip
+            try:
+                print('INFO:Copying resource as zip: %s...' % src_res.label())
+                copy_res_zip(src_res, dst_res, res_cache_dir)
+                return
+            except Exception:
+                try:
+                    print('INFO: second attempt to copy resource as zip: %s...'
```

Comment:
Why there are 2 attempts to upload the bundled contents of `resource` using zipped files but solely one when following a file-by-file approach?

Reply:
I would say for the same reason as we have introduced `STUBBORN` in other parts of `pyxnat`.

Reply:
odd but ok